File: apps/ops-browser/middleware.ts
import { NextResponse } from 'next/server';
import type { NextRequest } from 'next/server';

export function middleware(request: NextRequest) {
  const response = NextResponse.next();
  
  // Only add CSP in production (dev mode needs unsafe-inline/unsafe-eval for Next.js)
  const isDev = process.env.NODE_ENV === 'development';
  
  if (!isDev) {
    // Content Security Policy (production only)
    const apiBase = process.env.NEXT_PUBLIC_API_BASE || 'http://localhost:8000';
    const apiOrigin = new URL(apiBase).origin;
    
    response.headers.set(
      'Content-Security-Policy',
      `default-src 'self'; ` +
      `connect-src 'self' ${apiOrigin}; ` +
      `img-src 'self' data:; ` +
      `script-src 'self' 'unsafe-inline' 'unsafe-eval'; ` +  // Next.js needs these in production too
      `style-src 'self' 'unsafe-inline'; ` +  // Tailwind needs unsafe-inline
      `font-src 'self' data:;`
    );
  }
  
  // Security headers (always apply)
  response.headers.set('Referrer-Policy', 'strict-origin-when-cross-origin');
  response.headers.set('X-Content-Type-Options', 'nosniff');
  response.headers.set('X-Frame-Options', 'DENY');
  response.headers.set('X-XSS-Protection', '1; mode=block');
  
  // HSTS (if HTTPS)
  if (request.nextUrl.protocol === 'https:') {
    response.headers.set('Strict-Transport-Security', 'max-age=31536000; includeSubDomains');
  }
  
  return response;
}

export const config = {
  matcher: '/:path*',
};

File: apps/api/middleware.py
from starlette.middleware.base import BaseHTTPMiddleware


class SecurityHeadersMiddleware(BaseHTTPMiddleware):
    """
    Middleware to add security headers to all responses.
    """

    async def dispatch(self, request, call_next):
        response = await call_next(request)

        # Defense in Depth headers
        response.headers["X-Content-Type-Options"] = "nosniff"
        response.headers["X-Frame-Options"] = "DENY"
        response.headers["X-XSS-Protection"] = "1; mode=block"
        response.headers["Referrer-Policy"] = "strict-origin-when-cross-origin"

        # HSTS (Enable even for HTTP to ensure upgrade if proxy handles TLS)
        response.headers["Strict-Transport-Security"] = "max-age=31536000; includeSubDomains"

        # Content Security Policy (CSP)
        # Allows cdn.jsdelivr.net for Swagger UI assets
        csp = (
            "default-src 'self'; "
            "img-src 'self' data: https:; "
            "script-src 'self' 'unsafe-inline' 'unsafe-eval' https://cdn.jsdelivr.net; "
            "style-src 'self' 'unsafe-inline' https://cdn.jsdelivr.net"
        )
        response.headers["Content-Security-Policy"] = csp

        return response
